modules/fedimint-ln-client/src/pay.rs

```diff
@@ -268,49 +270,91 @@ impl LightningPayFunded {
         let contract_id = self.payload.contract_id;
         let timelock = self.timelock;
         let payment_hash = *common.invoice.payment_hash();
-        vec![StateTransition::new(
-            Self::gateway_pay_invoice(gateway, payload),
-            move |dbtx, result, old_state| {
-                Box::pin(Self::transition_outgoing_contract_execution(
-                    result,
-                    old_state,
-                    contract_id,
-                    timelock,
-                    dbtx,
-                    payment_hash,
-                    common.clone(),
-                ))
-            },
-        )]
+        let success_common = common.clone();
+        let timeout_common = common.clone();
+        let timeout_global_context = global_context.clone();
+        vec![
+            StateTransition::new(
+                Self::gateway_pay_invoice(gateway, payload),
+                move |dbtx, result, old_state| {
+                    Box::pin(Self::transition_outgoing_contract_execution(
+                        result,
+                        old_state,
+                        contract_id,
+                        dbtx,
+                        payment_hash,
+                        success_common.clone(),
+                    ))
+                },
+            ),
+            StateTransition::new(
+                await_contract_cancellable(contract_id, global_context.clone()),
+                move |dbtx, (), old_state| {
+                    Box::pin(try_refund_outgoing_contract(
+                        old_state,
+                        common.clone(),
+                        dbtx,
+                        global_context.clone(),
+                        format!("Gateway cancelled contract: {}", contract_id),
+                    ))
+                },
+            ),
+            StateTransition::new(
+                await_contract_timeout(timeout_global_context.clone(), timelock),
+                move |dbtx, (), old_state| {
+                    Box::pin(try_refund_outgoing_contract(
+                        old_state,
+                        timeout_common.clone(),
+                        dbtx,
+                        timeout_global_context.clone(),
+                        format!("Outgoing contract timed out, BlockHeight: {}", timelock),
+                    ))
+                },
+            ),
+        ]
     }
 
     async fn gateway_pay_invoice(
         gateway: LightningGateway,
         payload: PayInvoicePayload,
     ) -> Result<String, GatewayPayError> {
-        // Abort the payment if we can't reach the gateway within 30 seconds
-        // to prevent unexpected delays for the user.
-        let deadline = fedimint_core::time::now() + GATEWAY_API_TIMEOUT;
-
-        let mut last_error = None;
-        while fedimint_core::time::now() < deadline {
+        // Retry until the payment is successful. This future will be cancelled if
+        // the contract is cancelled.
+        loop {
```

Comment:
We don't necessarily want to retry for the entire timeout period. Chances are that the user figured out another way to pay by then.

If the timeout is a few days then it would be quite unexpected for the user that a transaction still goes through a day later or so. This can, in principle, happen anyway if the GW goes offline at the exactly wrong time, but will be very rare, this change makes it more likely imo.

Reply:
Right now we re-try every 30 seconds. Maybe we should expose a timeout parameter? It seems reasonable to me to wait a few minutes but I'm not sure if every wallet will want the same value. 

After the timeout crosses we could sleep until past the block timeout.

Reply:
The timeout path seems to be a separate state transition, so we could just do

```rust
std::future::pending().await;
```

here, which blocks forever.

Reply:
I added a timeout parameter of 180 seconds and changed the re-try interval to 10 seconds between retries. So if the gateway is failing unexpectedly, the client will try 18 times for a total of 3 minutes before giving up and blocking until the block timeout. Let me know if you think these guesses are unreasonable, I can change them.